Apply diff entries using async spi methods during bucket merge.

toregge opened the pull request:
@vekterli : please review
@baldersheim : FYI

Commit:
Apply diff entries using async spi methods during bucket merge.

vekterli commented on storage/src/vespa/storage/persistence/apply_bucket_diff_entry_result.h:
```diff
@@ -0,0 +1,30 @@
+    ApplyBucketDiffEntryResult(FutureResult future_result, spi::Bucket bucket, document::DocumentId doc_id, const char *op);
+    ApplyBucketDiffEntryResult(ApplyBucketDiffEntryResult &&rhs);
+    ~ApplyBucketDiffEntryResult();
+    void check_result();
```
Consider calling this something like `wait_and_check_result()` to indicate it also waits for the result to be available (or split into two functions)

vekterli commented on storage/src/vespa/storage/persistence/mergehandler.cpp:
```diff
@@ -617,9 +607,12 @@ MergeHandler::applyDiffLocally(
         byteCount += e._headerBlob.size() + e._bodyBlob.size();
     }
+    for (auto &result_to_check : async_results) {
+        result_to_check.check_result();
```
Should we split this into two phases (wait phase and check_result phase) to ensure all async ops have completed before potentially throwing an exception out of this method?

vekterli commented on storage/src/vespa/storage/persistence/apply_bucket_diff_entry_complete.cpp:
```diff
@@ -0,0 +1,35 @@
+        _result_handler->handle(*result);
+    }
+    _result_promise.set_value(std::move(result));
+    _latency_metric.addValue(_start_time.getElapsedTimeAsDouble());
```
Can `onComplete` for different document instances in the same merge be called from multiple threads concurrently? If so, this might race. Consider passing the elapsed time to the result future and aggregate this as part of the async result processing in the main merge handling thread.

vekterli reviewed (commented):
A couple of questions related to async completion (metric update concurrency and waiting for all replies) that I'd like addressed first. Otherwise LGTM!

Commit:
Move wait out of check_result.
Pass elapsed time along with result and update metric after having
checked the result.

vekterli reviewed (approved):
:+1:
